openstack: don't look up VMs by project ID.

oseiberts11 opened the pull request:
In a consul configuration, you can have this to discover the consul servers in an OpenStack project:
```
"retry_join": ["provider=os tag_key=role tag_value=consulserver"]
```
Most of the usual OS_* environment variables are used to authenticate to OpenStack. Unfortunately you can't specify the project NAME to authenticate with, when finding VMs with the given tag key and value. It only wants to use the project ID, and if it is not given, it tries a "somewhat ususal" (ahem) way to find it.

Apart from the authentication, the project ID was used when listing and filtering the VMs. There is no need to specify the project ID to find VMs. By default you only see the VMs of the project that you authenticated with anyway. (And there was no way in the discover code to auth with one project and look up another.) 

Finding the ID is annoying if you want to auth by project name, which was not even possible before.

Since we don't *need* to obtain the ID, we can get rid of the "somewhat ususal" (ahem) way to obtain it from the metadata service. When trying to use the metadata service, this assumes that you are running inside a VM in openstack, which is a pretty dubious assumption, and not our use case.

This merge request replaces #263 .

oseiberts11 committed:
openstack: don't look up VMs by project ID.

There is no need to specify the project ID to find VMs. By default you
only see the VMs of the project that you authenticated with anyway.
(And there was no way in the discover code to auth with one project and
look up another.) Finding the ID is annoying if you want to auth by
project name, which was not even possible before.
Since we don't *need* to obtain the ID, we can get rid of the somewhat
ususal way to obtain it from the metadata service.

oseiberts11 committed:
Merge branch 'master' into os-allow-auth-with-project-name

oseiberts11 commented:
Ping! Can somebody please look at this? One and a half years without attention gives a bad impression (sorry to say).

tgross reviewed (changes requested):
Hi @oseiberts11!

> One and a half years without attention gives a bad impression (sorry to say).

From the README:
> Note go-discover is considered feature complete and is currently maintained on an as-need basis for use with Vault and Consul _[ed: and Nomad!]_. Currently another solution go-netaddrs is being actively maintained, which provides a more flexible way of discovering ip addresses without having to tie it to specific infrastructure.

More generally, a lot of the Go libraries have recently changed hands internally, but these are pretty much in minimal maintenance mode except as needed for Vault/Nomad/Consul/etc. 

That being said, your PR handwaves over some of the context here as well.

> It tries a "somewhat ususal" (ahem) way to find it.

"(ahem)" what? This is probably clear to you but it isn't to me. Is this referring this?:

> When trying to use the metadata service, this assumes that you are running inside a VM in openstack, which is a pretty dubious assumption, and not our use case.

Why is this a dubious assumption? Why would you be using go-discover with an OpenStack configuration at all if you're not running on OpenStack VMs?

Do you have a reference link for the API to justify the change you're making? There's a bit of circular reasoning: you're saying you don't need the project ID for the List API so we don't need the project ID, because you're authenticated for a specific project anyways.

Also CI hasn't run here for some reason, so you'll probably need to rebase on tip and push again to kick that off. How have you tested this?

oseiberts11 committed:
Merge branch 'master' into os-allow-auth-with-project-name

oseiberts11 commented:
Note: the whole merge request is in the context where it is impractical to supply the UUID of a project ($OS_PROJECT_ID), but we can easily supply its name ($OS_PROJECT_NAME). If only we could practically supply the UUID, we would not have noticed this bug.

> Note go-discover is considered feature complete

That is fine; since this is about a bug (ignoring a standard OpenStack environment variable when it could easily be used), it would be maintenance and not adding a feature.

> Currently another solution go-netaddrs is being actively maintained, which provides a more flexible way of discovering ip addresses without having to tie it to specific infrastructure.

I guess we will have a look if that would fit with our use case.

> "(ahem)" what? This is probably clear to you but it isn't to me.

I was trying to remain polite and refrain from saying that looking at the metadata service to find a project UUID is a totally ludicrous and outrageously horrible way of doing things. There is no reason to believe that the owner of the VM is guaranteed to be the same as the openstack user we are authenticating with.

> Why is this a dubious assumption? 

Suppose you have a consul cluster that spans a network across VMs and non-VMs, where the consul servers to connect to are all inside VMs. In that case you won't have a metadata service available on the non-VMs.

Our case is actually a bit more subtle (I had to look it up, after all that time): we have set up a network that spans VMs owned by multiple projects. Non-admin projects can't see resources or metadata from other projects normally, but for networks you can set up some sharing. However, if you're searching for VMs based on tags and their values, you can see only your own VMs. If you attempt to find a project UUID from the metadata service, you get the value that corresponds to the owner of the VM. But that is NOT the project which we want to use to authenticate to OpenStack. As a result, the wanted VMs with consul servers are not found.

In all OpenStack applications, you can supply either the project UUID, or the name (and domain, if needed). In the second case, the UUID is looked up automatically. This is a totally normal use case, and it is a bug if it doesn't work like that.

> There's a bit of circular reasoning:

I'm not sure but I suppose I was not explaining it clearly.

With the servers.List() API you can filter for various properties of the VMs you want to see listed. I am mainly familiar with the OpenStack API directly, and not so much with the implementation in Go as it is used here, of course. What I want to say is that there is no excuse for needing the project UUID for the servers.List() function. It can safely be eliminated from the call[1]. You also don't need it for authenticating. Therefore there is no need at all to try to find the project UUID, and the weird code to do it can safely be removed. (The UUID can probably retrieved from some Go API after authenticating, but since we don't need the project UUID in any case, I did not look into that).

[1] Since, as a non-admin user, you only get to see the VMs of your own project (the one you authenticated with). Filtering for it makes no difference. Even as an admin user, you need to specify a special argument to list VMs from other projects than the one you authenticated with (such as `all_tenants=True` in `https://api...../v2.1/servers/detail?deleted=False&all_tenants=True`) . (Admittedly, OpenStack is inconsistent with this: for some APIs admin gets information about all projects by default. Also they still have not fully replaced the term "tenant" with "project" everywhere). See the API spec at https://docs.openstack.org/api-ref/compute/#list-servers and click "details" on the right. "By default the servers are filtered using the project ID associated with the authenticated request." At the "project_id" filter it explicitly says that "This filter only works when the all_tenants filter is also specified."

If you're not bound to the owner of the VM, you can also authenticate as another user and project than the owner of the VM.  (In our use case: we authenticate as the owner project of the VMs we want to find)

> so you'll probably need to rebase on tip and push again to kick that off. 

I merged the main branch into this one (apparently that was done a year ago as well, so I did the same), which produced no complaints. I'm not sure what github means with "This branch is out-of-date with the base branch" since that seems not true: it offers some kind of merge commit, which is the very thing I just did myself.

> How have you tested this?

I have tested this by 
- building the `main` executable of go-discover (`go build cmd/discover/main.go`) and invoking it with some arguments and environment variables.
- having an essentially identical change on branch https://github.com/syseleven/go-discover/commits/syseleven-consul-v1.20.4/ , which is with consul 1.20.4, and using that in production.

oseiberts11 committed:
`go fmt ./...` adds a newline at the end.

tgross reviewed (changes requested):
> I was trying to remain polite and refrain from saying that looking at the metadata service to find a project UUID is a totally ludicrous and outrageously horrible way of doing things. There is no reason to believe that the owner of the VM is guaranteed to be the same as the openstack user we are authenticating with.

Using the instance metadata store for this case is fairly typical for system-level services like Nomad and Consul in environments like AWS. I recognize there's not 100% overlap, but I think was the reasoning here. 

In any case, we cannot simply remove that behavior, because of backwards compatibility. Users may be relying on it to be that way now. Seeing as how you're introducing a new argument, it should be safe to detect that argument and skip the project ID instance metadata lookup.

oseiberts11 committed:
Restore lookup of project ID in metadata as a fallback

if neither the project name nor the ID is given, either as an argument
or in the customary OpenStack environment variables.

oseiberts11 commented:
So I think you're saying that it would be acceptable if I re-add the lookup in the metadata service as a fallback. I can live with that.

I added a commit to re-add this as a fallback. I moved the place where this is done though. It makes more sense to do it when looking up the other authentication parameters too: you want this info when you're authenticating. Also I only do it if you supply neither the project name, nor the ID, since only one of the two is needed. 

Here is what it looks like, first with the usual authentication environment variables set, then with the project-related ones cleared:
```
~/git/go-discover$ ./main addrs provider=os tag_key=foo tag_value=bar
Registered providers: [aliyun aws azure digitalocean gce linode mdns os packet scaleway softlayer srv tencentcloud triton vsphere]
[DEBUG] discover: Using provider "os"
2026/07/27 17:37:57 [DEBUG] discover-os: Using tag_key=foo tag_value=bar
[DEBUG] discover-os: Authenticating...
[DEBUG] discover-os: Creating client...
[DEBUG] discover-os: Found ip addresses: []
```
(with the given parameters it will not find any addresses)

then with the project-related ones cleared:
```
~/git/go-discover$ OS_PROJECT_ID= OS_PROJECT_NAME= ./main addrs provider=os tag_key=foo tag_value=bar
Registered providers: [aliyun aws azure digitalocean gce linode mdns os packet scaleway softlayer srv tencentcloud triton vsphere]
[DEBUG] discover: Using provider "os"
2026/07/27 17:39:25 [DEBUG] discover-os: Using tag_key=foo tag_value=bar
[INFO] discover-os: ProjectID not provided. Looking up in metadata...
discover-os: Error asking metadata for project_id: Get "http://169.254.169.254/openstack/latest/meta_data.json": dial tcp 169.254.169.254:80: connect: host is down
```
(from not inside a VM).

(I will stay away from the inconsistency that for some providers, the parameter is called `tag_name` and for others `tag_key`...)

tgross reviewed (approved):
LGTM! Thanks @oseiberts11!

tgross commented on provider/os/os_discover.go:
```suggestion
// Copyright IBM Corp. 2017, 2026
```

tgross commented on provider/os/os_discover.go:
```diff
@@ -138,6 +129,17 @@ func newClient(args map[string]string, l *log.Logger) (*gophercloud.ServiceClien
 
+	if projectID == "" && projectName == "" { // Use the one on the instance if not provided either by parameter or env
+		l.Printf("[INFO] discover-os: ProjectID not provided. Looking up in metadata...")
+		var err error;
```
```suggestion
		var err error
```

tgross reviewed (changes requested):
Sorry @oseiberts11 I hadn't re-run the workflows when I approved, and those pointed out a couple minor items. I've marked those as suggestions.

oseiberts11 committed:
Small cleanups

- update copyright date
- remove unneeded semicolon.

tgross reviewed (approved):
LGTM. We've got one more major item in-flight for this library at the moment. Once that's been merged we'll tag the release and then Consul will be able to pick it up.
